core/src/sumeragi/view_change.rs

```diff
@@ -108,249 +68,87 @@ impl Proof {
+    fn verify_with_state(
         &self,
+        peers: &HashSet<PeerId>,
+        max_faults: usize,
         latest_block: &HashOf<VersionedCommittedBlock>,
-        latest_view_change: &HashOf<Self>,
-    ) -> bool {
-        &self.payload.latest_block == latest_block
-            && &self.payload.previous_proof == latest_view_change
-    }
-
-    /// The `Reason` of this view change. Why should topology change?
-    pub const fn reason(&self) -> &Reason {
-        &self.payload.reason
-    }
-
-    /// Signatures of peers who signed this proof - therefore voting for this view change.
-    pub const fn signatures(&self) -> &SignaturesOf<Self> {
-        &self.signatures
-    }
-}
-
-/// Payload of [`Proof`]
-#[derive(Debug, Clone, Decode, Encode, IntoSchema)]
-pub struct ProofPayload {
-    /// Hash os the previous view change proof.
-    previous_proof: HashOf<Proof>,
-    /// Latest committed block hash.
-    latest_block: HashOf<VersionedCommittedBlock>,
-    /// Reason for a view change.
-    reason: Reason,
-}
-
-/// Reason for a view change.
-#[derive(Debug, Display, Clone, Decode, Encode, FromVariant, IntoSchema)]
-pub enum Reason {
-    /// Proxy tail have not committed the block in time.
-    #[display(fmt = "Commit timeout")]
-    CommitTimeout(CommitTimeout),
-    /// Transaction was sent to the leader and no corresponding
-    /// receipt was received from the leader for it in time.
-    #[display(fmt = "Transaction receipt not received")]
-    NoTransactionReceiptReceived(NoTransactionReceiptReceived),
-    #[display(fmt = "Block creation timeout")]
-    /// Transaction reached leader but no block was created.
-    BlockCreationTimeout(Box<BlockCreationTimeout>),
-}
-
-/// Block `CommitTimeout` reason for a view change.
-#[derive(Debug, Clone, Copy, Decode, Encode, IntoSchema)]
-pub struct CommitTimeout {
-    /// The hash of the block in discussion in this round.
-    pub hash: HashOf<VersionedValidBlock>,
-}
-
-/// `NoTransactionReceiptReceived` (from leader) reason for a view change.
-#[derive(Clone, Debug, Encode, Decode, Copy, IntoSchema)]
-pub struct NoTransactionReceiptReceived;
-
-/// `BlockCreationTimeout` reason for a view change.
-#[derive(Copy, Clone, Debug, Encode, Decode, IntoSchema)]
-pub struct BlockCreationTimeout;
+    ) -> usize;
 
-/// A chain of view change proofs. Stored in block for roles to be known at that point in history.
-#[derive(Debug, Clone, Default, Decode, Encode, IntoSchema)]
-pub struct ProofChain {
-    proofs: Vec<Proof>,
+    fn insert_proof(
+        &mut self,
+        peers: &HashSet<PeerId>,
+        max_faults: usize,
+        latest_block: &HashOf<VersionedCommittedBlock>,
+        new_proof: &Proof,
+    );
 }
 
-impl ProofChain {
-    /// Initialize an empty proof chain.
-    #[must_use]
-    pub const fn empty() -> Self {
-        Self { proofs: Vec::new() }
-    }
-
-    /// Verify the view change proof chain.
-    pub fn verify_with_state(
+impl ProofChain for Vec<Proof> {
```

Comment:
OK. I see you found your legs w.r.t. traits.

Reply:
yes